### PR TITLE
Iter9

### PR DESCRIPTION
**Pull request by 25x8**

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -1,46 +1,121 @@
...
 
 // SaveCounterMetric - сохраняет метрику типа counter
 func (s *MemStorage) SaveCounterMetric(name string, value int64) error {
 	s.Lock()
-	defer s.Unlock()
 	s.counters[name] += value
+	s.Unlock()
+
+	if s.storeInterval == 0 {
+		if err := s.SaveToFile(); err != nil {
```
Comment (arttet) [inline cmd/server/main.go:114]: [CHANGE REQUEST] Please follow the SOLID principles

File: cmd/server/main.go [hunk 2]
```diff
@@ -81,6 +156,76 @@ func (s *MemStorage) GetAllMetrics() map[string]interface{} {
...
+		return nil
+	}
+
+	file, err := os.Open(s.filePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	decoder := json.NewDecoder(file)
```
Comment (arttet) [inline cmd/server/main.go:205]: [CHANGE REQUEST] Please follow the SOLID principles.

File: cmd/server/main.go [hunk 3]
```diff
@@ -111,11 +256,58 @@ func handleGetValue(s *MemStorage) http.HandlerFunc {
 	}
 }
 
+func handleGetValueJSON(s *MemStorage) http.HandlerFunc {
```
Comment (arttet) [inline cmd/server/main.go:259]: [CHANGE REQUEST] Please split your application by Tiers.
https://medium.com/@shrestha.matina.20/what-is-the-3-tier-architecture-4520522e0720

**Review by arttet — CHANGES_REQUESTED** [review]

File: cmd/server/main.go [hunk 4]
```diff
@@ -179,29 +371,142 @@ func handleUpdateMetric(s *MemStorage) http.HandlerFunc {
 
 		w.WriteHeader(http.StatusOK)
 		fmt.Fprintf(w, "Metric %s updated", metricName)
+	})
+}
+
+func handleUpdateMetricJSON(s *MemStorage) http.HandlerFunc {
```
[the remark this answers is not in the document]
Comment (25x8) [reply cmd/server/main.go:377]: HERE WE ARE USE MIDDLEWARES, SO IT CAN'T BE JUST METHODS

File: cmd/server/main.go [hunk 5]
```diff
@@ -2,206 +2,108 @@ package main
...
-	r := mux.NewRouter()
+	defer logger.Sync()
+
+	h := handler.Handler{Storage: newStorage}
 
 	// Маршруты для обновления метрик и получения их значений
-	r.HandleFunc("/update/{type}/{name}/{value}", handleUpdateMetric(storage)).Methods(http.MethodPost)
-	r.HandleFunc("/value/{type}/{name}", handleGetValue(storage)).Methods(http.MethodGet)
-	r.HandleFunc("/", handleGetAllMetrics(storage)).Methods(http.MethodGet)
+	r.Handle("/update/{type}/{name}/{value}", middleware.GzipMiddleware(logger.RequestLogger(http.HandlerFunc(h.HandleUpdateMetric)))).Methods(http.MethodPost)
```
Comment (arttet) [inline cmd/server/main.go:99]: [CHANGE REQUEST] Please simplify the `main` package.

File: internal/logger/logger.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	"net/http"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+// Log будет доступен всему коду как синглтон.
+// Никакой код навыка, кроме функции Initialize, не должен модифицировать эту переменную.
+// По умолчанию установлен no-op-логер, который не выводит никаких сообщений.
+var Log *zap.Logger = zap.NewNop()
```
Comment (arttet) [inline internal/logger/logger.go:13]: [CHANGE REQUEST] Please don't use global variables.

File: internal/storage/metrics.go [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package storage
+
+type Metrics struct {
```
Comment (arttet) [inline internal/storage/metrics.go:3]: [CLARIFICATION] Why is it here?
[CLARIFICATION] What is it? DTO or the domain object?

**Review by arttet — CHANGES_REQUESTED** [review]

File: cmd/agent/senders/http_sender.go [hunk 8]
```diff
@@ -32,17 +34,44 @@ func (s *HTTPSender) Send(metrics map[string]interface{}) error {
...
-		req, err := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(nil))
+
+		// Сжатие тела запроса (в данном случае тело пустое, но это пригодится для реальных данных)
+		var compressedBody bytes.Buffer
+		gzipWriter := gzip.NewWriter(&compressedBody)
+		_, err := gzipWriter.Write([]byte{}) // Пустое тело для текущего примера
+		if err != nil {
+			return err
+		}
+		gzipWriter.Close()
```
Comment (arttet) [inline cmd/agent/senders/http_sender.go:45]: [CHANGE REQUEST] Please review the Go layouts again.
https://github.com/golang-standards/project-layout/blob/master/cmd/README.md

File: cmd/app/app.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+		var err error
+		restore, err = strconv.ParseBool(envRestore)
+		if err != nil {
+			log.Fatalf("Invalid RESTORE value: %v", err)
+		}
+	}
+
+	// Инициализация логгера
+	if err := logger.Initialize("info"); err != nil {
+		panic(err)
```
Comment (arttet) [inline cmd/app/app.go:64]: [CHANGE REQUEST] Please don't panic
https://github.com/uber-go/guide/blob/master/style.md#dont-panic

File: cmd/app/app.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+	// Обработка сигнала завершения для сохранения метрик
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		if err := storage.SaveToFile(newStorage, fileStoragePath); err != nil {
+			log.Printf("Error saving metrics on shutdown: %v", err)
+		}
+		os.Exit(0)
```
Comment (arttet) [inline cmd/app/app.go:91]: [SUGGESTION] This is not a graceful shutdown.

File: internal/storage/mem_storage.go [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+		allMetrics[name] = value
+	}
+	for name, value := range s.counters {
+		allMetrics[name] = value
+	}
+	return allMetrics
+}
+
+// SaveToFile - сохраняет метрики в файл
+func SaveToFile(s *MemStorage, filePath string) error {
```
Comment (arttet) [inline internal/storage/mem_storage.go:85]: [SUGGESTION] So, you can move it to another package.

**Review by arttet — APPROVED** [review]

**Pull request merged**